### PR TITLE
Add feature to flash information for the current page load only

### LINKED ISSUE #2: Support flash "now" parameter, for saving information for current requests
`kirby-flash` works really well when I want to save a success message or an error for the next page load, and I return a redirect from a controller.

However, I'd like to use `kirby-flash` in situations where I render a 422 response immediately after a form error. In this case, I'd like the flash error message to show up in the response for the current request (and not the subsequent page load).

[Rails has a similar pattern](https://spartchou.gitbooks.io/ruby-on-rails-basic/content/syntax/flash_vs_flashnow.html), and the flash object has a [`now` method](https://api.rubyonrails.org/v6.1.0/classes/ActionDispatch/Flash/FlashHash.html#method-i-now) to support this use case.

Would you be open to this feature?

I think the cleanest API would probably be to support a 3rd, optional boolean argument called "now" on `flash->set()` and the `flash` helper. Under the hood, this would store the data in a separate temporary array on the `Flash` instance, which would not be persisted to the session. `flash` and `flash->get()` would read from both the session data and the temporary data when looking for a key.

### PR DESCRIPTION
**Pull request by s3ththompson**
Sometimes it can be useful to flash a message for the current page load, not the next one. This use case arises when a message should be shown in a response to the same request, rather than a redirect.

I extended the helper and `set` with the following signature: `set($key, $value = '', $now = false)`. Under the hood, this essentially just skips setting the value to the Kirby session.

---

In the process of adding this feature, I also added tests to test the behavior of subsequent page loads. I did this by creating a test helper method called `nextPageLoad` which clears the Flash static `$instance` property.

These new tests revealed an existing bug: setting any new flash data carried all messages for the current page load over to the next one. I fixed this bug, but the corrected behavior might technically be considered a breaking change.

### CODE REVIEW

**Comment by s3ththompson** [comment]
Resolves #2

File: README.md [hunk 1]
```diff
@@ -8,8 +8,8 @@ Allows you to "flash" data to the session, which will be available via the sessi
 
 Very useful for:
 
-- Saving submitted form data for form validation, specifically for the [Post/Redirect/Get](https://en.wikipedia.org/wiki/Post/Redirect/Get) design pattern
-- Displaying Success or Error messages after a page reload
+-   Saving submitted form data for form validation, specifically for the [Post/Redirect/Get](https://en.wikipedia.org/wiki/Post/Redirect/Get) design pattern
+-   Displaying Success or Error messages after a page reload
```
Comment (mzur) [inline README.md]: Please revert.

File: README.md [hunk 2]
```diff
@@ -21,8 +21,8 @@ Elsewhere...
 
 ```php
 <?php if (flash('thanks_message')): ?>
-    <?php echo flash('thanks_message') ?>
-<?php endif ?>
+    <?php echo flash('thanks_message'); ?>
+<?php endif; ?>
```
Comment (mzur) [inline README.md]: Please revert (I prefer this without `;`).

File: README.md [hunk 3]
```diff
@@ -71,9 +70,9 @@ flash('messages.errors'); // Array( 0 => 'Email is required', 1 => 'Password is
 <div class="alert alert-error">
     <?php foreach (flash('messages.errors') as $message): ?>
         <div><?= html($message) ?></div>
-    <?php endforeach ?>
+    <?php endforeach; ?>
 </div>
-<?php endif ?>
+<?php endif; ?>
```
Comment (mzur) [inline README.md]: Please revert.

File: composer.json [hunk 4]
```diff
@@ -30,9 +34,9 @@
     "require-dev": {
         "phpunit/phpunit": "^6.0",
         "mzur/kirby-defuse-session": "^1.0",
-        "getkirby/kirby": "^3.0"
+        "getkirby/cms": "^3.0"
     },
-    "extra": {
-        "kirby-cms-path": "vendor/getkirby/kirby"
+    "require": {
+        "getkirby/composer-installer": "^1.2"
```
Comment (mzur) [inline composer.json]: Let's keep the changes to this file to the other PR please.

File: tests/TestCase.php [hunk 5]
```diff
@@ -3,15 +3,31 @@
...
      * Default preparation for each test.
      */
-    public function setUp()
+    public function setUp(): void
     {
         parent::setUp();
         Defuse::defuse();
+        $this->nextPageLoad();
+        // run two full page loads to ensure each test is isolated
+        $this->nextPageLoad();
```
Comment (mzur) [inline tests/TestCase.php:19]: Why are two required?

**Review by mzur — CHANGES_REQUESTED** [review]
Thanks for the suggestion! So as I understand this change is basically for convenience of coding? Otherwise I don't understand why you want to "flash" something if you don't actually want to flash it to the session.

You said that "now" comes from Rails but I'm not sure I like the term. If `now` is `false` the value is still flashed "now". Maybe call it "temporary" or "volatile"?

File: README.md (hunk 1 above)
Comment (s3ththompson) [reply README.md]: oops, I had an overzealous code formatter on. Reverting all stylistic changes right now.

File: tests/TestCase.php (hunk 5 above)
Comment (s3ththompson) [reply tests/TestCase.php:19]: This ensures each test is fully isolated.

The first `nextPageLoad` is the one that is actually supported by `flash` (after all the whole point is to remember data across a page load), the second `nextPageLoad` simulates an extra page load *after* the flash.

```
               +-----------+            +-----------+
               | PageLoad1 |            | PageLoad2 |
               +-----------+            +-----------+
                     |                        |
                     |                        |
-------------------\ |                        |
| Data is flashed. |-|                        |
|------------------| |                        |
                     |                        |
                     | Data can be read       |
                     |----------------------->|
                     |                        | -----------------------------\
                     |                        |-| Data is no longer readable |
                     |                        | |----------------------------|
                     |                        |
```

**Comment by s3ththompson** [comment]
First off, thanks for reviewing @mzur! I've resolved the stylistic changes.

My use case is actually not about convenience... it has to do with using `kirby-flash` as part of `kirby-form`. If this PR lands, I'd like to make a followup PR on `kirby-form` allowing the ability to pass this flag through.

The use case has to do with conventions for processing form validation. If a user posts a form with data that doesn't validate, I'd like to return a `422 Unprocessable Entity` response code, with the form and errors in the response body.

```php
// site/controllers/my-page.php
use Jevets\Kirby\Form;

return function ($kirby) {
    $form = new Form([
        'name' => [
            'rules'     => ['required'],
            'message'   => ['Name is required']
        ],
        'phone' => [],
    ]);

    if ($kirby->request()->is('POST')) {
        // if form does not validate,
        // return 422 with errors in response body

        // pass `now: true` so that errors are rendered for *this* response body,
        // but *not* the subsequent page navigation
        if (!$form->validates(now: true)) {
          kirby()
            ->response()
            ->code(422);
          return compact('form');
        } else { // else, return 302 to success page
          return go('/success');
        }
    }

    return compact('form');
};
```

This pattern is the form validation [convention in Rails](https://guides.rubyonrails.org/getting_started.html#:~:text=422%20Unprocessable%20Entity) and it's necessary for getting things to work with the framework-agnostic clientside routing library [Turbo](https://turbo.hotwired.dev/). See, e.g. their guidance on form errors [here](https://turbo.hotwired.dev/handbook/drive#redirecting-after-a-form-submission).

It looks like pattern contrasts with the original convention of `kirby-form` (which seems similar to that of [Laravel](https://laravel.com/docs/9.x/validation#quick-displaying-the-validation-errors)), in which if form data doesn't validate, a 302 redirect is returned and the form errors are shown on the _subsequent_ page load.

Either way, hopefully you're not opposed to supporting this pattern indirectly via the `now` flag (or `volatile`, etc.), which would allow `kirby-flash` and `kirby-form` to be used more flexibly, depending on the user's preference (and accommodating frontend libraries which may expect one or the other pattern).

Thanks for considering!

**Comment by mzur** [comment]
I see your point, thanks for the explanation. I'd like to propose a different approach. Rather than defining that a flashed variable should be removed in the `set()` method, we could implement the `pull()` method as an alternative to the `get()` method in the Flash class (and a `flash('key', pull: true)` argument for the helper). The `pull()` method returns the flashed value and then removes it from the session. This is inspired by Laravel again.

With `kirby-form` this could be used as an additional constructor argument: `new Form(pull: true)`. The remaining methods `errors()`, `error()` etc. can be used as usual. What do you think?

**Comment by s3ththompson** [comment]
Thanks for the reply!

Hmm... in my opinion, the benefit of deciding the lifecycle of the flashed variable at the time you call `set()` is that you have the most context about the intent of the message at that point. For example in the above example, if I am returning error messages in the response I might want to flash something for the current request, but if I am redirecting to a success message I might want to flash something for the next request.

```php
// site/controllers/my-page.php
return function ($kirby) {
    if ($kirby->request()->is('POST')) {
        if ($error) {
          flash('key', $error,  now: true);
          kirby()
            ->response()
            ->code(422);
          return []; // render the page
        } else {
          flash('key', 'Success');
          return go('/success'); // redirect
        }
    }

    return [];
};
``` 

The other challenge with relying on `pull` is that it breaks the idea that the `flash` global helper is idempotent for reading data for the duration of a request. For example, I might want to read the key in multiple snippets:

```html
<h1>Site Title</h1>
<?php if (flash('key') === 'Success' ): ?>
    <div class="green-icon"></div>
<?php else: ?>
    <div class="red-icon"></div>
<?php endif; ?>
``` 

```html
<div class="message"><?= flash('key') ?></div>
```

This wouldn't be possible if I had to use `flash('key', pull: true)` to keep my message from carrying over into the next request.

FWIW it looks like Laravel actually [supports a `now` method](https://laravel.com/docs/9.x/session#flash-data) for flashing data for the current request. They mention the same use case in that commit (flashing data without redirecting): https://github.com/laravel/framework/commit/2aecef2e625be2e93bc05874feaf24041a8eafab.

Thanks for your patience discussing this. I'm definitely open to an alternate API, but I don't think this feature would be as usable with a `pull`-based approach.

File: README.md [hunk 6]
```diff
@@ -107,6 +107,18 @@ flash('messages.errors', ['Email is required.', 'Phone is required.']);
...
+
+Sometimes it can be useful to flash a message for the current page load, not the next one. This use case
+arises when a message should be shown in a response to the same request, rather than a redirect.
+
+The `flash()` helper method can be called with an optional **third parameter**, a boolean to toggle whether to flash _now_ instead of in the next page load. The default value of this parameter is `false`: `flash($key, $value = '', $now = false)`.
+
+```php
+flash('message', 'Message for redirect'); // flash on the next page load (default)
+flash('message', 'Message for this response', true); // flash in a response to the same request
+```
```
Comment (mzur) [inline README.md]: I find this phrasing to be misleading. The flashing is done *now* in both cases. Only with `$now = true` it doesn't persist.
```suggestion
The `flash()` helper method can be called with an optional third parameter, a boolean to toggle whether to keep the flashed variable only for the current request. The default value of this parameter is `false` which will keep the flashed variable for the next request.

```php
flash('message', 'Message for redirect'); // Keep for next request
flash('message', 'Message for this response', true); // Keep only for current request
```
```

File: src/Flash.php [hunk 7]
```diff
@@ -25,7 +25,7 @@ class Flash
      *
      * @var array
      */
-    protected $data;
+    protected $initialData;
```
Comment (mzur) [inline src/Flash.php]: I don't see a need to rename the variable. Any specific reason?

File: src/helpers.php [hunk 8]
```diff
@@ -12,13 +12,14 @@
      *
      * @param string $key
      * @param mixed optional $value to set
+     * @param mixed optional set $value for current page load only
```
Comment (mzur) [inline src/helpers.php]:
```suggestion
     * @param boolean optional set $value for current page load only
```

File: tests/FlashTest.php [hunk 9]
```diff
@@ -15,11 +14,13 @@ public function testGetInstance()
 
     public function testSessionKey()
     {
+        $flash = Flash::getInstance();
```
Comment (mzur) [inline tests/FlashTest.php]: Why is this needed?

File: tests/FlashTest.php [hunk 10]
```diff
@@ -15,11 +14,13 @@ public function testGetInstance()
...
         $this->assertEquals('_flash', Flash::sessionKey());
     }
 
     public function testSetSessionKey()
     {
+        $flash = Flash::getInstance();
```
Comment (mzur) [inline tests/FlashTest.php]: Why is this needed?

File: tests/FlashTest.php [hunk 11]
```diff
@@ -48,4 +49,39 @@ public function testInstances()
...
+    {
+        $flash = Flash::getInstance();
+        $flash->set('key', 'value');
+        $this->nextPageLoad();
+        $flash = Flash::getInstance();
+        $this->assertEquals('value', $flash->get('key'));
+        $flash->set('key2', 'value2');
+        $this->nextPageLoad();
+        $flash = Flash::getInstance();
+        $this->assertNull($flash->get('key'));
```
Comment (mzur) [inline tests/FlashTest.php:73]: Maybe check if `key2` is still there?
```suggestion
        $this->assertNull($flash->get('key'));
        $this->assertEquals('value2', $flash->get('key2'));
```

**Review by mzur — CHANGES_REQUESTED** [review]
That's fair! On to me being nitpicky :smile:

File: tests/FlashTest.php (hunk 9 above)
Comment (s3ththompson) [reply tests/FlashTest.php]: good catch! fixed in latest commit

**Comment by s3ththompson** [comment]
@mzur great suggestions, I think I addressed all of the nits. Thanks!

**Comment by s3ththompson** [comment]
If this PR is merged, I'd be happy to open a PR against `kirby-form` to support passing through the `now` parameter.

**Review by mzur — APPROVED** [review]
All good, thanks!

**Review by mzur — CHANGES_REQUESTED** [review]
Actually, no. The new tests show that the changes to `composer.json` cause issues. Can you please fix/revert that?

**Comment by s3ththompson** [comment]
oops, i don't know how the wrong Kirby package name snuck in there

**Review by mzur — APPROVED** [review]

**Comment by mzur** [comment]
Looks like I've been to hasty again. I just noticed that the changes to `composer.json` here seem to basically revert your PR #1. Can you please check if this is all intended? I will release a new version once this is clarified.

**Comment by s3ththompson** [comment]
~@mzur the net of #1 was that the package name should be [`getkirby/cms`](https://packagist.org/packages/getkirby/cms). I'm not sure how this PR ended up with the old package name along the way, but it's correct now, everywhere. Thanks for your patience!~

**Comment by s3ththompson** [comment]
Yikes. Good catch, you're right. Should we revert the merge, or should I fix in a new PR?

**Comment by mzur** [comment]
I think it's easiest with a new PR.
